accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,577 @@
+- Azure prototyping.
+  - The Azure team needs a writable DOM, and supporting `dynamic` is important for some scenarios but not all. A [prototype](https://docs.microsoft.com/en-us/dotnet/api/azure.core.dynamicjson?view=azure-dotnet-preview) has been created. Work is also being coordinated with the C# team to support Intellisense with `dynamic`, although that is not expected to be implemented in time to be used in 6.0.
+
+# Dynamic
+The dynamic feature in C# is enabled by both language features and implementation in the `System.Linq.Expressions` assembly. The core interface detected by the language to enable this support is [IDynamicMetaObjectProvider](https://docs.microsoft.com/en-us/dotnet/api/system.dynamic.idynamicmetaobjectprovider?view=net-5.0).
+
+There are implementations of `IDynamicMetaObjectProvider` including [DynamicObject](https://docs.microsoft.com/en-us/dotnet/api/system.dynamic.dynamicobject?view=net-5.0) and [ExpandObject](https://docs.microsoft.com/en-us/dotnet/api/system.dynamic.expandoobject?view=netcore-3.1). However, both implementations are not ideal:
+- There are many public members used for wiring up dynamic support, but are of little to no value for consumers so they would be confusing when mixed in with other methods intended to be used for writable DOM support.
+- It is not possible to efficiently combine writable DOM classes that are not tied to `dynamic` and `System.Linq.Expressions`:
+  - If combined, there would need to be an assembly reference to the very large `System.Linq.Expressions.dll` event when `dynamic` features are not needed.
+  - If separate, the `dynamic` classes delegate to separate writable DOM classes, which doubles the object instances and increases the concept count around interop between the types returned from `dynamic` vs. the writable DOM classes.
+- `ExpandoObject` does not support case insensitivity and throws an exception when accessing missing properties (a `null` is desired instead).
+- `DynamicObject` prevents potential optimizations including property-lookup.
+
+Thus, the design presented here for `dynamic` assumes:
+- Implementation of `IDynamicMetaObjectProvider`.
+- Use of the writable DOM without requiring a reference to `System.Linq.Expressions.dll` (when `dynamic` support is not desired).
+- Efficient implementation of tying both `IDynamicMetaObjectProvider` and the writable DOM classes into a single instance. Newtonsoft also has a single object model with [JToken](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Linq_JToken.htm) which implements `IDynamicMetaObjectProvider`.
+
+## Adding a System.Text.Json.Dynamic.dll (STJD) assembly
```

Comment:
I understand that there might be technical limitations that we won't be able to overcome but can we start with the best-case scenario where dynamic support doesn't require additional dependency/configuration code and see if we can achieve it?

Reply:
Perhaps. This is also a problem we have today for other assembly references in STJ that we don't have a solution for today. Consider:
- STJ has a reference to SLE because of the new `System.Object` converter's implementation of `IDynamicMetaObjectProvider`.
- To function properly, at run-time that converter must be added before (de)serialization occurs, so the reference to SLE is there normally.
- To allow the linker to remove the reference to SLE, the converter shouldn't be registered\added by default, thus an "opt-in" so the linker can walk dependencies.
- The linker then would also need to consider applications (e.g. stand-alone and Blazor) which it doesn't do today (it only considers framework assemblies) and would need to change STJ based on the application and whether the opt-in was called.
  - The linker would remove the opt-in method since it is not called.
  - Then the linker would remove the custom converter since the opt-in method is no longer there.
  - Then the linker would remove the reference to SLE since it is no longer needed.

@eerhardt are aware of plans in 6.0 to support application-to-framework dependencies and "assembly reference removal" like above? Is it feasible?

Reply:
I'm not sure I fully understand the question. But even in 5.0:

* If System.Text.Json exposes a method that is never invoked by the app, the ILLinker will trim the method.
* If no methods/types in an assembly are used (i.e. SLE), the the ILLinker will remove the whole assembly and all references to it.

Reply:
Thanks @eerhardt I changed the doc to assume ILLinker instead of a separate assembly. I was unable to prototype however due to missing 6.0 packages at this time.